File: python/637-Average of Levels in Binary Tree.py
from typing import List
from bintree_helper import *

class Solution:
    def averageOfLevels(self, root) -> List[float]:
        "DFS, Time: O(n), Space: O(n)"
        if root is None: return [0]
        sums = []
        counts = []
        def dfs(node, level):
            if len(sums) == level: 
                sums.append(0)
                counts.append(0)
            sums[level] += node.val 
            counts[level] += 1
            if node.left: dfs(node.left, level+1)
            if node.right: dfs(node.right, level+1)
        dfs(root, 0)
        return [s/c for s, c in zip(sums, counts)]
    def averageOfLevels(self, root) -> List[float]:
        "BFS, Time: O(n), Space: O(n)"
        from collections import deque
        if root is None: return [0]

        q = deque([root])
        result = []
        while q:
            n = len(q)
            s, count = 0, 0
            for _ in range(n):
                node = q.popleft()
                s += node.val
                count += 1
                if node.left: q.append(node.left)
                if node.right: q.append(node.right)
            result.append(s/count)
        return result
root = deserialize("[3,9,20,null,null,15,7]")
result = Solution().averageOfLevels(root)
assert result == [3, 14.5, 11]

File: cpp/637. Average of Levels in Binary Tree.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;
typedef vector<double> vd;


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
// BFS, Time: O(n), Space: O(n)
public:
    vector<double> averageOfLevels(TreeNode* root) {
        if(!root) return vd({0});
        queue<TreeNode*> q; q.push(root);
        vd ans;
        while(!q.empty()) {
            ll total = 0;
            int sz = q.size();
            for(int i = 0; i < sz; i++) {
                TreeNode *node = q.front(); q.pop();
                total += node->val;
                if(node->left) q.push(node->left);
                if(node->right) q.push(node->right);
            }
            ans.push_back((double)total/sz);
        }
        return ans;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    
}